### PR TITLE
barebones sql shell

### PR DESCRIPTION
**Pull request by mberhault**
Invoke with:
$ ./cockroach sql --addr=localhost:10000
Sample use:

```
> show databases
Database
foo
> use foo
0 rows affected
> show index from baz
Table Name  Unique  Seq Column
baz primary true  1 k
> select * from baz
k v
a z
b c
> quit
```

This is insanely hacky and should not be merged,
it's merely useful to poke around the sql interface.

### CODE REVIEW

File: cli/sql.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+			osExit(1)
+		}
+		words := strings.Split(line, " ")
+		if len(words) == 0 {
+			continue
+		}
+
+		// This is a really terrible way of figuring out how to handle
+		// the input.
+		command := strings.ToUpper(words[0])
```
Comment (petermattis) [inline cli/sql.go]: The `psql` shell uses back-slash to separate "shell commands" from SQL:

```
pmattis=# \?
General
  \copyright             show PostgreSQL usage and distribution terms
  \g [FILE] or ;         execute query (and send results to file or |pipe)
  \gset [PREFIX]         execute query and store results in psql variables
  \h [NAME]              help on syntax of SQL commands, * for all commands
  \q                     quit psql
  \watch [SEC]           execute query every SEC seconds
```

Comment (MBerhault) [inline cli/sql.go]: Ah. I'm afraid I'm only familiar with the mysql shell.
Still, this is the least of the issues in this thing. I've been poking around for a decent shell library (good input/output, multi-line support, history, proper signal handling, and maybe even auto-completion), but I'm not having much luck. If we want this kind of thing in the future, we may have to roll our own, this was mostly for my own personal amusement.

File: cli/sql.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+			// Format all rows using tabwriter.
+			tw := new(tabwriter.Writer)
+			tw.Init(term, 0, 8, 0, '\t', 0)
+			fmt.Fprintf(tw, "%s\n", strings.Join(cols, "\t"))
+			for rows.Next() {
+				strs := make([]string, len(cols))
+				vals := make([]interface{}, len(cols))
+				for i := range vals {
+					vals[i] = &strs[i]
+				}
```
Comment (petermattis) [inline cli/sql.go]: More efficient to allocate `strs` and `vals` once outside the loop.

File: cli/sql.go (hunk 1 above)
Comment (petermattis) [inline cli/sql.go]: We'll almost certainly want to support this at some point. Would be good in the short term (i.e. even before the shell) to support a `sql -c <SQL>` command to immediately run the specified SQL.

Comment (MBerhault) [inline cli/sql.go]: I started off with cmdline statements, but ultimately I found having to constantly retype things like "use XX" to be really tedious.

Comment (petermattis) [inline cli/sql.go]: We should be able to specify the database to use via a flag (or as part of a URL).

Comment (bdarnell) [inline cli/sql.go]: At least in mysql, "use" is never required; you can always access tables by their fully-qualified name (and you can also pass the database name on the command line).

I'd rather support "quit" and "exit" as normal-looking commands instead of requiring something like `\q`; we don't need a separate namespace unless we start to have a lot of them.

**Comment by bdarnell** [comment]
LGTM. I'd go ahead and merge this - it's pretty primitive but will still be useful in its current form.

File: cli/sql.go (hunk 1 above)
Comment (MBerhault) [inline cli/sql.go]: and using <db>.<table> does work in all the commands I've tried (create/select/show/insert). sql already has a way to specify the database that people are familiar with, I'd rather not add another flag.

**Comment by MBerhault** [comment]
ok. I'll clean this up a bit and merge.

File: cli/sql.go (hunk 2 above)
Comment (MBerhault) [inline cli/sql.go]: Done.

File: cli/sql.go (hunk 1 above)
Comment (petermattis) [inline cli/sql.go]: How do you specify the database in SQL? If you're referring to `USE`, that is MySQL specific. And `USE` doesn't even work properly with connection pools such as the one in `database/sql`, because `USE` specifies the database for a connection but you have no way of knowing when a new connection is created. Specification of the database is usually database/driver specific.

Comment (petermattis) [inline cli/sql.go]: Btw, both `psql` and `mysql` support specifying the database as a flag (`psql -d` and `mysql -D`).

Comment (MBerhault) [inline cli/sql.go]: ok, even adding one of those, we currently have no way of specifying the database in the url. We'd have to extract the database name from the url and pass that around to pre-fill the request. We probably do want to support it at that point, but it seems outside the scope of this toy shell.

Comment (petermattis) [inline cli/sql.go]: I'm not suggesting this needs to be done in this PR, but that it should be the roadmap. Yes, we can't currently specify the database in the URL, but we should fix that. Both the mysql and postgres golang drivers allow specifying the database that way (albeit via different formats).

File: cli/sql.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	Long: `
+Open a sql shell running against the cockroach database at --addr.
+`,
+	Run: runTerm,
+}
+
+// processOneLine takes a line from the terminal,
+// determines what type of statement it is, runs it,
+// and displays the result.
+// TODO(marc): handle multi-line, this will require ';' terminated statements.
```
Comment (petermattis) [inline cli/sql.go]: You should definitely require `;` terminated statements. I'm pretty sure this is what other SQL shells do. They keep reading lines until they see one that is terminated with a `;`.

Comment (tschottdorf) [inline cli/sql.go]: MySQL also has `\G` (or something like that) which alters the display format of the result (from a table to "one list per row")

File: cli/sql.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+	// This is a really terrible way of figuring out how to handle
+	// the input.
+	command := strings.ToUpper(words[0])
+	switch command {
+	case "EXIT", "QUIT":
+		return true, nil
+	case "SELECT", "SHOW":
+		// Issues a query and examine returned Rows.
+		rows, err := db.Query(line)
```
Comment (petermattis) [inline cli/sql.go]: Btw, there is no real harm in calling `db.Query` for all SQL statements. The same code is called in the driver for `db.Query` and `db.Exec`, the only difference is the return type (the rows or a count of the rows).

File: cli/sql.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	words := strings.Split(line, " ")
+	if len(words) == 0 {
+		return false, nil
+	}
+
+	// This is a really terrible way of figuring out how to handle
+	// the input.
+	command := strings.ToUpper(words[0])
+	switch command {
+	case "EXIT", "QUIT":
```
Comment (petermattis) [inline cli/sql.go]: Does `control-D` work?

**Comment by petermattis** [comment]
LGTM

File: cli/sql.go (hunk 4 above)
Comment (MBerhault) [inline cli/sql.go]: True, especially since we don't properly set RowsAffected, so it always displays zero.

File: cli/sql.go (hunk 5 above)
Comment (MBerhault) [inline cli/sql.go]: Yup.

File: cli/sql.go (hunk 3 above)
Comment (MBerhault) [inline cli/sql.go]: agreed with ';'. I can do a basic "has trailing ;", looking for unescaped ones is trickier (and annoying).
as for special toggles (eg: \G), that's way overboard for this.

Comment (petermattis) [inline cli/sql.go]: You could use the SQL parser for parsing. Not quite sure how that would work, but something to keep in mind.

File: cli/sql.go (hunk 4 above)
Comment (MBerhault) [inline cli/sql.go]: switched to always using Query, just displaying "OK" when there are no columns.
